lang/rust/avro/src/error.rs

```diff
@@ -478,6 +478,9 @@ pub enum Error {
 
     #[error("Invalid Avro data! Cannot read codec type from value that is not Value::Bytes.")]
     BadCodecMetadata,
+
+    #[error("Schemas are not compatible. '{0}'")]
+    CompatibilityError(String),
```

Comment:
Should this be a separate error? I don't think it's wise to keep adding errors to this `Error`. 

As a user, if you try to deserialize and fails, you shouldn't have to handle this error.

```rs
match Schema::parse_str(reader_raw_schema) {
    CompatibilityError => unreachable!()
}
```

I think compatibility should have its own enum with a set of errors so the user can react to them properly.
```rs
#[derive(Error, Debug)]
pub enum CompatibilityError {
  #[error("We got {input}, but {output} was expected")]
  WrongKind { input: SchemaKind, output: SchemaKind},
  
  #[error("Reader's symbols must contain all writer's symbols")]
  MissingSymbols,

    #[error("Could not calculate compatibility")]
    Inconclusive
}
```

And more...



Thoughts @martin-g @marcosschroh ?

Reply:
Currently, AvroResult is defined as
```rust
pub type AvroResult<T> = Result<T, Error>;
```
with Error as this Enum, so you can't derive.
Would be possible with
```rust
#[derive(thiserror::Error, Debug)]
pub enum CompatibilityError {
```
But, then, you could not build an AvroResult (until AvroResult is redefined as  `Result<T, thiserror::Error>`, but i don't know what would be other consequences
WDYT @martin-g ?

Reply:
Yeah, that would also have to change. For the compatibility, you can just use a `Result<(), CompatibilityError`. Looking at the code, it was just a `bool` before. I would avoid from the get go tying it to `AvroResult`.

Reply:
In a separate commit I have added the new error `CompatibilityError` with its types so the users can specify business logic per error type. Some errors type are missing because the current code is not covering uses cases like `logicalTypes` and`enum orders`,  we will have to add them in the future as we continue using this library.